### ISSUE TITLE
Add an API endpoint that resolves arbitrary URLs into specs

### ISSUE THREAD

**Issue by yuvipanda** [jupyterhub/binderhub#322]
Riffing off the awesomebar concept that @choldgraf and @betatim (and maybe @mariusvniekerk?) have been talking about, I think it'll be useful to have an API endpoint that takes an arbitrary URL and returns a JSON response with possible spec info. 

So if you give it a GitHub url (long or short), it'll return the fact that it's a GitHub URL and a spec that can be fed into the rest of the API. Same for gist, or the specific git URLs, or gitlab (or future providers).

Responding with a JSON response allows clients to provide appropriate UI for this functionality before starting builds. In the mybinder.org UI, we can change an icon or something in the bar as soon as URL is pasted and before launch is clicked.

We can implement this with a 'detect' method on each of the providers that get passed this URL and can return True + a spec...

**Comment by yuvipanda** [comment]
This would also keep provider code all python rather than split it into Python + JS, at the cost of an extra network request. However, if we're going to support other UI providers, making this be a API endpoint that can be hit rather than JS code to be included sounds great.

**Comment by rgbkrk** [comment]
> I think it'll be useful to have an API endpoint that takes an arbitrary URL and returns a JSON response with possible spec info.

👍

**Comment by mariusvniekerk** [comment]
if @rgbrk uses the build url directly, we could use it to fully validate the build parts of the new providers without having to have a UI yet.  :)
